src/org/jitsi/meet/test/ConnectionTimeTest.java

```diff
@@ -0,0 +1,388 @@
+            {
+                data[s.ordinal()][i] 
+                    = s.execute(participant);
+            }
+            
+        }
+        
+        if(data[TimeMeasurements.CONNECTION_ATTACHED.ordinal()][0] == null
+            && data[TimeMeasurements.CONNECTION_CONNECTED.ordinal()][0] == null)
+        {
+            // conference failed
+            fail("Conference failed!");
+            return;
+        }
+        else if(data[TimeMeasurements.CONNECTION_ATTACHED.ordinal()][0] != null)
+        {
+            connectingTime = TimeMeasurements.CONNECTION_ATTACHING;
+            connectedTime = TimeMeasurements.CONNECTION_ATTACHED;
+        }
+        else if(data[TimeMeasurements.CONNECTION_CONNECTED.ordinal()][0] != null)
+        {
+            connectingTime = TimeMeasurements.CONNECTION_CONNECTING;
+            connectedTime = TimeMeasurements.CONNECTION_CONNECTED;
+        }
+    }
+    
+    /**
+     * Refreshes the second participant.
+     */
+    private static void refreshSecondParticipant()
+    {
+        ConferenceFixture.close(participant);
+        participant = ConferenceFixture.startSecondParticipant();
+    }
+    
+    /**
+     * Waits for VIDEO_RENDER time measurement to be calculated. That way
+     * we can be sure that all other measurements are performed already.
+     */
+    private static void waitForVideoRender()
+    {
+        TestUtils.waitForCondition(participant, 10, new ExpectedCondition<Boolean>()
+        {
+
+            @Override
+            public Boolean apply(WebDriver w)
+            {
+                return TimeMeasurements.VIDEO_RENDER.execute(participant) != null;
+            }
+        });
+    }
+    
+    /**
+     * Evaluates passed time measurement value. Fails if the difference between
+     * the previous measurement and the current one is bigger than the 
+     * threshold for the current measurement. 
+     * @param s the time measurement that will be evaluated.
+     */
+    private void checkTime(TimeMeasurements s)
+    {
+        if(s == TimeMeasurements.MUC_JOINED)
+        {
+            checkThreshold(data[connectedTime.ordinal()], s);
+        }
+        else
+        {
+            Double[] prevStepData = s.getPrevStep() == null ? null : 
+                data[s.getPrevStep().ordinal()];
+            checkThreshold(prevStepData, s);
+        }
+    }
+    
+    /**
+     * Compares the threshold for the passed time measurement with the 
+     * median of the subtracted values of the data from the passed time 
+     * measurement and the passed array. 
+     * @param previousStepTimes array with times that will be subtract
+     * @param s the time measurement
+     */
+    private void checkThreshold(Double[] previousStepTimes, TimeMeasurements s)
+    {
+        Double[] difference = (previousStepTimes == null)?  data[s.ordinal()] :
+            subtractArrays(previousStepTimes, data[s.ordinal()]);
+        assertTrue(getMedian(difference) < s.getThreshold());
```

Comment:
We need to add here some meaningful message what had actually failed, if it fail.